src/experimental/patterns/ec2-app.ts

```diff
@@ -0,0 +1,132 @@
+import { Duration } from "aws-cdk-lib";
+import type { CfnAutoScalingGroup } from "aws-cdk-lib/aws-autoscaling";
+import { UpdatePolicy } from "aws-cdk-lib/aws-autoscaling";
+import { Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
+import type { GuStack } from "../../constructs/core";
+import type { GuEc2AppProps } from "../../patterns";
+import { GuEc2App } from "../../patterns";
+
+export interface GuEc2AppExperimentalProps extends Omit<GuEc2AppProps, "updatePolicy"> {}
+
+/**
+ * An experimental pattern to instantiate an EC2 application that is updated entirely via CloudFormation.
+ *
+ * NOTE: The "autoscaling" deployment type in Riff-Raff is not valid with this pattern.
+ * Please remove it from any manually created `riff-raff.yaml` file.
+ *
+ * This pattern sets the update policy of the `AWS::AutoScaling::AutoScalingGroup` to `AutoScalingReplacingUpdate`.
+ * When a CloudFormation update is applied, a SECOND autoscaling group is created alongside the current one.
+ * Once all instances in the new ASG are healthy in the target group, CloudFormation will remove the current ASG,
+ * leaving one ASG provisioned.
+ *
+ * This pattern also updates the User Data, running some commands AFTER yours.
+ * These changes are wrapped in start and end marking comments.
+ *
+ * This pattern should improve the reliability of scaling events triggered during a deployment.
+ * Unlike in Riff-Raff's "autoscaling" deployment, scaling alarms are never suspended.
+ * TODO test scaling alarm behaviour.
+ *
+ * To update the application's code, a CloudFormation update must be triggered.
+ * The best way to do this is to include the build number in the application artifact.
+ * TODO test User Data includes a build number.
+ *
+ * NOTE: This pattern:
+ *  - Is NOT compatible with the "autoscaling" Riff-Raff deployment type.
+ *  - Your application should include a build number in its filename.
+ *    This value will change across builds, and therefore create a CloudFormation template difference to be deployed.
```

Comment:
If the AMI also gets updated by CloudFormation, would that be a separate CloudFormation change, or would we try and do it together in one update? I think the reason this is important is that CloudFormation will rollback failed changes, and we wouldn't it to rollback halfway to the old ASG with the new AMI.

Reply:
Currently, Riff-Raff attempts to update the AMI on each deployment. That is, the changes are bundled together into one.